### ISSUE TITLE
[ SUGGESTION ] - Multiple Docker hosts

### ISSUE THREAD

**Issue by Its4Nik** [sergi0g/cup#13]
i would really love a feature where we can add remote docker sockets just like in dozzle if you know that.

I am not sure on how this could work because i am not a backend developer :/

This would allow user to monitor remote docker / podman sockets as well

**Comment by sergi0g** [comment]
Thanks for the suggestion! I'll look into it!

It may be a bit tricky to implement because of the library currently used. I might also need your help for testing.

**Comment by Its4Nik** [comment]
> Thanks for the suggestion! I'll look into it!
> 
> It may be a bit tricky to implement because of the library currently used. I might also need your help for testing.

I can definitely help with testing, since I have like 6 Servers to try it on to really stress this tool :D

**Comment by jdevera** [comment]
Super interested in this and want to comment on a setup.

I use https://github.com/Tecnativa/docker-socket-proxy on every docker host so that I can use some tools like https://gethomepage.dev/ to report status without having write access.

I believe starting with supporting accessing docker information through the HTTP interface even if only one host is used is a great first step in the direction of supporting multiple hosts.

**Comment by 004a** [comment]
As an alternative, you could take input from a yaml or toml config file with a list of the images, diun style. Then you wouldn't need access to the socket at all and could support checking for more images than what's running on the local or any connected host.

**Comment by sergi0g** [comment]
Hey @Its4Nik! I've looked into this issue a bit and from what I see, projects that support multiple hosts run an agent on each host and communicate with that, without exposing any sockets on the network (this is done by Dozzle and Portainer).

If that were to happen, a very simple way to achieve this functionality would be to run a Cup instance on each server and add an option to the config file to add more hosts. Then, the "main" instance could just use the API endpoint from each server and aggregate the results. What do you think?

**Comment by sergi0g** [comment]
Hey @Its4Nik! I've looked into this issue a bit and from what I see, projects that support multiple hosts run an agent on each host and communicate with that, without exposing any sockets on the network (this is done by Dozzle and Portainer).

If that were to happen, a very simple way to achieve this functionality would be to run a Cup instance on each server and add an option to the config file to add more hosts. Then, the "main" instance could just use the API endpoint from each server and aggregate the results. What do you think?

**Comment by Its4Nik** [comment]
> Hey @Its4Nik! I've looked into this issue a bit and from what I see, projects that support multiple hosts run an agent on each host and communicate with that, without exposing any sockets on the network (this is done by Dozzle and Portainer).
> 
> If that were to happen, a very simple way to achieve this functionality would be to run a Cup instance on each server and add an option to the config file to add more hosts. Then, the "main" instance could just use the API endpoint from each server and aggregate the results. What do you think?

Yeah that could work but (i am not sure how cup works under the hood) is it possible to add an env var to toggle the frontend on/off? I guess that would improve container reliability and more usecases in low power environments

Just as a question if the frontend actually is already very light on the host

**Comment by sergi0g** [comment]
I could add a config option to turn the frontend off, however it would all code for it would still be present in the binary. It's just a few hundred kb.

**Comment by Its4Nik** [comment]
@sergi0g Yeah I think that would be nice :D

**Comment by Its4Nik** [comment]
@sergi0g do you have any way of contacting you privately? I have some general questions, not _really_ related to cup.

**Comment by sergi0g** [comment]
@Its4Nik you can send me an email at s.backboard587@passinbox.com

**Comment by zmweske** [comment]
I would be more than happy to help test as well. Also, it appears that Dozzle, homepage, and What'sUpDocker can all access the docker socket via a tcp port, which is all that https://github.com/Tecnativa/docker-socket-proxy exposes. Sometimes it is denoted by just host and port and other times it is denoted as `tcp://10.0.10.1:2375`. 

Either way, I have been looking for a project like this for so long!! Thank you so much for all your work!!!

**Comment by sergi0g** [comment]
Hey!

I haven't had the time to implement this feature yet, but it's planned for version 3 which I'm working on at the moment. I'm kind of stuck on semver support, so it'll be a while until it's ready (I have a _very_ busy schedule, I'm sorry). I'll tell you when it's ready to test.

Thanks for the idea of using something else instead of the docker socket. I know the benefits and I'll try to implement that too if I can!

**Comment by sergi0g** [comment]
Hello everyone!

I'm ready to start working on this feature and I'd like your feedback about something.

Other projects that offer multiple host support with "agents" typically distribute separate images for the "main" host with the web frontend and the "agents". That shouldn't be too hard to implement for Cup. However I was thinking that some users might like to run 2 (or more) servers in web mode and have each server connect to the other, so no matter which dashboard they viewed, they'd see the same update list. This could also enable agents having their own agents. This mode of operation would require careful design and complex code, so I'm wondering if any of you would be interested in such a feature or not.

Do you think it's worth implementing?

**Comment by zmweske** [comment]
I would agree with @jdevera - start with one web server that is able to connect to multiple docker sockets. This has several benefits:
1. Reduced complexity on your codebase
2. reuse other existing and separately maintained projects (utilize what they have done while reducing your maintenance workload) like https://github.com/Tecnativa/docker-socket-proxy - this project can expose your docker socket in a secure way (limited access, read-only, etc) so that you can safely connect to the TCP docker socket over the network
3. it reduces the footprint of your codebase which reduces the risk of a security issue
4. it prevents you from potentially putting a lot of work into something that there is already an adequate solution to (granted, I get the appeal of tackling new problems like this)
This is not to say that it could be evaluated in the future, but I think it would be best to start with something simple and go from there.

**Comment by sergi0g** [comment]
Thanks for the feedback! I'll go with the simpler approach for now, as you suggested. Users will have to make sure to not create a two-way connection between servers, to avoid problems. Hopefully, you'll have a nightly version to test in a next few days.

**Comment by sergi0g** [comment]
Bumping this thread because I haven't received a reply. I'll assume you forgot.

**Comment by Its4Nik** [comment]
> Bumping this thread because I haven't received a reply. I'll assume you forgot.

Guilty... But I am going to try it this evening probably :)

**Comment by zmweske** [comment]
Finally have time to get back to the homelab following some family issues, trying to check it out now! Sorry for the delay but excited to check it out!

**Comment by sergi0g** [comment]
Hello! Finally somebody responded 😄! I'll push another nightly so you can test the latest features. Hopefully the release is coming up in a few days. Sorry for not having proper documentation. It's ready but I don't know how to deploy it alongside the existing docs. Maybe I need to research this... 😂
Anyway, if you can't get it working you can make an issue and I'll help!

**Comment by zmweske** [comment]
Idk what I did, maybe just simply removed the docker image, but then I re-ran the existing config and everything worked, so idk what was going on, I guess some stale cache or something...? Either way it seems to be working now!! Looks great!!

**Comment by sergi0g** [comment]
OK, I'm glad you got it working! V3 will be a major update, since the API has changed almost completely (you can check how much data is returned yourself) and the config file schema has also completely changed. Don't worry, I'll create a fully-detailed change log describing everything that's changed.

**Comment by jdevera** [comment]
Hey! Thank you for this update and apologies for not coming back to test this sooner. I got the multiple server feature working and I'm super happy seeing all statuses together!

WRT to the refresh policy, I think the first one you described is the simplest and it is also effective. I recommend waiting for potential performance problems to be experienced or reported before complicating development to avoid them.


I love the collapsible section for the other servers in the UI and would love it even more if the first section with the local images was also collapsible.

Thanks again for making these changes!
